include: "../../include_map_to_protein/Snakefile"
include: "../../include_map_to_RNA/Snakefile"

assert "MAX_RAM_MB" in config, "'MAX_RAM_MB' not in config"
assert "MAX_THREADS" in config, "'MAX_THREADS' not in config"
assert "MAX_VRAM_MB" in config, "'MAX_VRAM_MB' not in config"
MAX_RAM_MB = config["MAX_RAM_MB"]
MAX_THREADS = config["MAX_THREADS"]
MAX_VRAM_MB = config["MAX_VRAM_MB"]

SEQS = []

class DenovoSequencing():
    def __init__(self, 
                project, #name of your project
                MS_data, #path to the folder with MS data (.mzML)
                model='casanovo_massivekb.ckpt', #name of your Casanovo model file
                map_to_protein=False, #whether to map to protein databas
                extra_fasta=None,
                canonical_proteome=None,
                map_to_RNA=False,
                genome='GRCh38.p14', #name of your genome that you generated with IndexGenome.jar
                extra=None, #Fasta file containing extra sequences (DNA will be 6-frame translated!)!
                rnaseq=None, #Fasta file containing extra transcript sequences (will be 3-frame translated!)!
                var=None #File containing variants (each line); could also be in ${input%.csv}.var, in which case you do not have to specify this parameter!
                ):
        self.project = project 
        self.MS_data = MS_data 
        self.model = model
        self.map_to_protein = map_to_protein
        self.extra_fasta = extra_fasta
        self.canonical_proteome = canonical_proteome
        self.map_to_RNA = map_to_RNA
        self.genome = genome
        self.extra = extra
        self.rnaseq = rnaseq
        self.var = var

        if self.map_to_protein:
            PROTEIN_MAPPINGS.append(ProteinMapping(self.project, self.canonical_proteome, self.extra_fasta))
        if self.map_to_RNA:
            RNA_MAPPINGS.append(RNAMapping(self.project, self.genome, self.extra, self.rnaseq, self.var))

if "todo" in config:
    include: f"profiles/{config['todo']}/todo.py"

def get_dataset(project):
    for dataset in SEQS:
        if dataset.project == project:
            return dataset
    raise ValueError(f"Dataset {project} not found")

rule all:
    input:
        [f"_data/casanovo/{seq.project}/{seq.project}.mztab" for seq in SEQS] +
        [f"_data/ahocorasick/{protein_mapping.project}/{protein_mapping.project}_mapped.csv" for protein_mapping in PROTEIN_MAPPINGS] +
        [f"_data/prism/{rna_mapping.project}/{rna_mapping.project}_toprism.csv.pep.annotated.csv.gz" for rna_mapping in RNA_MAPPINGS]

rule run_Casanovo:
    input:
        MS_data = lambda wildcards: get_dataset(wildcards.project).MS_data
    output:
        mztab = "_data/casanovo/{project}/{project}.mztab"
    singularity:
        "workflow/containers/image/python.3.10.15.gpu.ubuntu.sif"
    params:
        tmp_mztab = "_data/casanovo/{project}/{project}",
        model = lambda wildcards: get_dataset(wildcards.project).model,
    resources:
        vram=MAX_VRAM_MB,
        threads=MAX_THREADS,
        mem_mb=MAX_RAM_MB,
        runtime=600
    shell:
        '''
        casanovo sequence -f -o {params.tmp_mztab} -m model/{params.model} -c model/config.yaml -v info {input.MS_data}/*.mzML
        mv {params.tmp_mztab} {output.mztab}
        '''



#rule postprocessing:



# rule have_fun:
#               __________ 
#              |          |
#              | ~@**#~~  |
#        ,,,   |  ________|
#      (O___0) \/  
#   \    | |    /
#    \ \     \ /
#      /     /
#      |     |
